frame/support/procedural/src/construct_runtime/expand/unsigned.rs

```diff
@@ -0,0 +1,74 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+use crate::construct_runtime::Pallet;
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::Ident;
+
+pub fn expand_outer_validate_unsigned(
+	runtime: &Ident,
+	pallet_decls: &[Pallet],
+	scrate: &TokenStream,
+	use_v2: bool,
+) -> TokenStream {
+	let mut pallet_names = Vec::new();
+	let mut query_validate_unsigned_part_macros = Vec::new();
+
+	for pallet_decl in pallet_decls {
+		if pallet_decl.exists_part("ValidateUnsigned") {
+			let name = &pallet_decl.name;
+
+			pallet_names.push(name);
+
+			if use_v2 {
+				let path = &pallet_decl.pallet;
+				query_validate_unsigned_part_macros.push(quote!( #path::__is_validate_unsigned_part_defined!(#name); ));
+			}
+		}
+	}
+
+	quote! {
+		#( #query_validate_unsigned_part_macros )*
+
+		impl #scrate::unsigned::ValidateUnsigned for #runtime {
+			type Call = Call;
+
+			fn pre_dispatch(call: &Self::Call) -> Result<(), #scrate::unsigned::TransactionValidityError> {
+				#[allow(unreachable_patterns)]
+				match call {
+					#( Call::#pallet_names(inner_call) => #pallet_names::pre_dispatch(inner_call), )*
+					// pre-dispatch should not stop inherent extrinsics, validation should prevent
+					// including arbitrary (non-inherent) extrinsics to blocks.
+					_ => Ok(()),
+				}
+			}
+
+			fn validate_unsigned(
+				#[allow(unused_variables)]
+				source: #scrate::unsigned::TransactionSource,
+				call: &Self::Call,
+			) -> #scrate::unsigned::TransactionValidity {
+				#[allow(unreachable_patterns)]
+				match call {
+					#( Call::#pallet_names(inner_call) => #pallet_names::validate_unsigned(source, inner_call), )*
+					_ => #scrate::unsigned::UnknownTransaction::NoUnsignedValidator.into(),
+				}
+			}
+		}
```

Comment:
I know there wasn't test for it in frame/support/src/validate_unsigned, but a small test for them in frame/support/test/test/construct_runtime.rs would be better.